File: cmd/git-ai-commit/message.go
package main

import (
	"errors"
	"fmt"
	"os"
	"strings"
	"unicode/utf8"
)

const maxSubjectRunes = 72

func sanitizeMessage(message string) string {
	message = strings.ReplaceAll(message, "\r\n", "\n")
	message = strings.TrimSpace(message)
	message = strings.TrimPrefix(message, "```")
	message = strings.TrimSuffix(message, "```")
	message = strings.TrimSpace(message)

	lines := strings.Split(message, "\n")
	cleaned := make([]string, 0, len(lines))
	for _, line := range lines {
		if strings.HasPrefix(strings.TrimSpace(line), "```") {
			continue
		}
		cleaned = append(cleaned, strings.TrimRight(line, " \t"))
	}

	return strings.TrimSpace(strings.Join(cleaned, "\n"))
}

func validateMessage(message string) error {
	if message == "" {
		return errors.New("chat completion returned an empty message")
	}

	subject := firstLine(message)
	if utf8.RuneCountInString(subject) > maxSubjectRunes {
		return fmt.Errorf("generated subject exceeds %d characters", maxSubjectRunes)
	}

	return nil
}

func writeMessageFile(path, message string) error {
	existing, err := os.ReadFile(path)
	if err != nil && !errors.Is(err, os.ErrNotExist) {
		return err
	}

	content := message + "\n"
	if trimmed := strings.TrimSpace(string(existing)); trimmed != "" {
		content += "\n" + strings.TrimLeft(string(existing), "\n")
	}

	return os.WriteFile(path, []byte(content), 0o644)
}

func trimToUTF8Bytes(input string, maxBytes int) string {
	if maxBytes <= 0 {
		return ""
	}
	if len(input) <= maxBytes {
		return input
	}

	trimmed := input[:maxBytes]
	for !utf8.ValidString(trimmed) && len(trimmed) > 0 {
		trimmed = trimmed[:len(trimmed)-1]
	}
	return trimmed
}

func trimWithNoticeAtLineBoundary(input string, maxBytes int, notice string) (string, bool) {
	if maxBytes <= 0 {
		return "", len(input) > 0
	}
	if len(input) <= maxBytes {
		return input, false
	}

	available := maxBytes - len(notice)
	if available <= 0 {
		return trimToUTF8Bytes(notice, maxBytes), true
	}

	trimmed := trimToUTF8Bytes(input, available)
	if idx := strings.LastIndex(trimmed, "\n"); idx > 0 {
		trimmed = trimmed[:idx+1]
	}
	if trimmed == "" {
		trimmed = trimToUTF8Bytes(input, available)
	}
	if trimmed == "" {
		return trimToUTF8Bytes(notice, maxBytes), true
	}

	return trimmed + notice, true
}

func firstLine(input string) string {
	if idx := strings.IndexByte(input, '\n'); idx >= 0 {
		return input[:idx]
	}
	return input
}

File: src/message.rs
use crate::text_budget;

pub const MAX_SUBJECT_CHARS: usize = 72;

pub fn sanitize_message(message: &str) -> String {
    let normalized = message.replace("\r\n", "\n");
    let trimmed = normalized
        .trim()
        .trim_start_matches("```")
        .trim_end_matches("```");

    let sanitized = trimmed
        .trim()
        .lines()
        .filter(|line| !line.trim_start().starts_with("```"))
        .map(|line| line.trim_end_matches([' ', '\t']))
        .collect::<Vec<_>>()
        .join("\n")
        .trim()
        .to_string();

    enforce_subject_limit(&sanitized)
}

pub fn validate_message(message: &str) -> Result<(), String> {
    if message.trim().is_empty() {
        return Err("model returned an empty message".to_string());
    }

    let subject_len = first_line(message).chars().count();
    if subject_len > MAX_SUBJECT_CHARS {
        return Err(format!(
            "generated subject exceeds {MAX_SUBJECT_CHARS} characters"
        ));
    }

    Ok(())
}

pub fn trim_to_utf8_bytes(input: &str, max_bytes: usize) -> String {
    if max_bytes == 0 {
        return String::new();
    }
    if input.len() <= max_bytes {
        return input.to_string();
    }

    let mut end = max_bytes;
    while !input.is_char_boundary(end) {
        end -= 1;
    }
    input[..end].to_string()
}

pub fn trim_with_notice_at_line_boundary(
    input: &str,
    max_bytes: usize,
    notice: &str,
) -> (String, bool) {
    text_budget::trim_with_notice_at_line_boundary(
        input,
        max_bytes,
        notice,
        |value| value.len(),
        trim_to_utf8_bytes,
    )
}

pub fn first_line(input: &str) -> &str {
    input.split('\n').next().unwrap_or(input)
}

fn enforce_subject_limit(message: &str) -> String {
    if message.trim().is_empty() {
        return String::new();
    }

    let mut lines = message.lines().map(str::to_string).collect::<Vec<_>>();
    if let Some(subject) = lines.first_mut()
        && subject.chars().count() > MAX_SUBJECT_CHARS
    {
        *subject = truncate_subject(subject);
    }

    lines.join("\n").trim().to_string()
}

fn truncate_subject(subject: &str) -> String {
    if subject.chars().count() <= MAX_SUBJECT_CHARS {
        return subject.to_string();
    }

    let mut cutoff = 0;
    let mut last_space_cutoff = None;

    for (idx, ch) in subject.char_indices() {
        let next = idx + ch.len_utf8();
        if subject[..next].chars().count() > MAX_SUBJECT_CHARS {
            break;
        }
        cutoff = next;
        if ch.is_whitespace() {
            last_space_cutoff = Some(idx);
        }
    }

    let preferred = last_space_cutoff.unwrap_or(cutoff);
    subject[..preferred].trim_end().to_string()
}

#[cfg(test)]
mod tests {
    use super::{
        MAX_SUBJECT_CHARS, sanitize_message, trim_to_utf8_bytes, trim_with_notice_at_line_boundary,
    };

    const DIFF_HUNK_TRUNCATED_NOTICE: &str = "[hunk truncated]\n";

    #[test]
    fn keeps_utf8_valid_when_trimming() {
        let trimmed = trim_to_utf8_bytes("你好abc", 5);
        assert_eq!(trimmed, "你");
        assert!(trimmed.is_char_boundary(trimmed.len()));
    }

    #[test]
    fn prefers_line_boundaries() {
        let input = "line-1\nline-2\nline-3\nline-4\nline-5\n";
        let max = "line-1\nline-2\n".len() + DIFF_HUNK_TRUNCATED_NOTICE.len();
        let (trimmed, truncated) =
            trim_with_notice_at_line_boundary(input, max, DIFF_HUNK_TRUNCATED_NOTICE);

        assert!(truncated);
        assert!(trimmed.starts_with("line-1\nline-2\n"));
        assert!(trimmed.ends_with(DIFF_HUNK_TRUNCATED_NOTICE));
    }

    #[test]
    fn falls_back_to_notice_for_tiny_budget() {
        let (trimmed, truncated) =
            trim_with_notice_at_line_boundary("abcdef", 4, DIFF_HUNK_TRUNCATED_NOTICE);

        assert!(truncated);
        assert!(!trimmed.is_empty());
        assert!(trimmed.len() <= 4);
    }

    #[test]
    fn removes_code_fences() {
        let sanitized = sanitize_message("```text\nfeat: add tests\n\nbody\n```");
        assert!(!sanitized.contains("```"));
        assert!(sanitized.contains("feat: add tests"));
    }

    #[test]
    fn truncates_overlong_subject_at_word_boundary() {
        let sanitized = sanitize_message(
            "feat: add OpenRouter model context auto-detection and provider debug logging\n\n- body",
        );
        assert!(super::first_line(&sanitized).chars().count() <= MAX_SUBJECT_CHARS);
        assert_eq!(
            super::first_line(&sanitized),
            "feat: add OpenRouter model context auto-detection and provider"
        );
        assert!(sanitized.contains("\n\n- body"));
    }
}
